acme_diags/driver/qbo_driver.py

```diff
@@ -0,0 +1,178 @@
+        xclim[month] = np.nanmean(xraw[month::months_per_year])
+    num_years = int(np.floor(len(x_deseasoned) / months_per_year))
+    # Iterate through all years in x_deseasoned (same number as in xraw)
+    for year in np.arange(num_years):
+        year_index = year * months_per_year
+        # Iterate through all months of the year
+        for month in np.arange(months_per_year):
+            month_index = year_index + month
+            # Subtract the month's mean over num_years from xraw's data for this month in this year
+            # i.e., get the difference between this month's value and it's "usual" value
+            x_deseasoned[month_index] = xraw[month_index] - xclim[month]
+    return x_deseasoned
+
+
+def get_20to40month_fft_amplitude(qboN, levelN):
+    # Calculates the amplitude of wind variations in the 20 - 40 month period
+    pspecN = np.zeros(levelN.shape)
+    ampN = np.zeros(levelN.shape)
+
+    for ilev in np.arange(len(levelN)):
+        # `qboN[:, ilev]` returns the entire 0th dimension for ilev in the 1st dimension of the array.
+        y_input = deseason(np.squeeze(qboN[:, ilev]))
+        Y = scipy.fftpack.fft(y_input)
+        n = len(Y)
+        # TODO: f is frequency?
+        f = np.arange(n / 2) / n
+        period = 1 / f
+        values = Y[0:int(np.floor(n / 2))]
+        fyy = values * np.conj(values)
+        # Choose the range 20 - 40 months that captures most QBOs (in nature)
+        pspecN[ilev] = 2 * np.nansum(fyy[(period <= 40) & (period >= 20)])
+        ampN[ilev] = np.sqrt(2 * pspecN[ilev]) * (f[1] - f[0])
+    return pspecN, ampN
+
+
+def process_u_for_power_spectral_density(data_region):
+    # Average over vertical levels and horizontal area
+    levBot = 22
+    levTop = 18
+    # Average over lat and lon
+    data_lat_lon_average = cdutil.averager(data_region, axis='xy')
+    # Average over vertical
+    x0 = np.nanmean(np.array(data_lat_lon_average(level=(levTop, levBot))), axis=1)
+    # x0 should now be 1D
+    return x0
+
+
+def ceil_log2(x):
+    """
+    Given a number, calculate the exponent for the next power of 2.
+
+    Example:
+        ceil_log2(16) = 4
+        ceil_log2(17) = 5
+    """
+    return np.ceil(np.log2(x)).astype('int')
+
+
+def get_psd_from_deseason(xraw, periodNew):
+    x0 = deseason(xraw)
+
+    # Sampling frequency: assumes frequency of sampling = 1 month
+    Fs0 = 1
+    T0 = 1 / Fs0
+    L0 = len(xraw)
+    t0 = np.arange(0, L0) * T0
+    NFFT0 = 2 ** ceil_log2(L0)
+
+    # Apply fft on x0 with n = NFFT
+    X0 = scipy.fftpack.fft(x0, n=NFFT0) / L0
+    # Frequency (cycles/month)
+    f0 = Fs0 * np.arange(0, (NFFT0 / 2 + 1)) / NFFT0
+
+    period0 = 1 / f0
+
+    # Amplitude
+    AMP0 = 2 * abs(X0[0:int(NFFT0 / 2 + 1)])
+    # Power spectral density
+    PSDx0 = AMP0 ** 2 / L0
+    Tend0 = T0 * L0
+    # Total spectral power
+    Pxf0 = Tend0 * np.sum(PSDx0)
+    # TODO: Because why?
```

Comment:
Why do we need to flip?

Reply:
The spectral density and amplitudes were calculated as a function of frequency, but we later want to perform an interpolation using the period (interpolate values of AMP0_flipped & PSDx0_flipped from period0_flipped to periodNew) and we want to make sure that period is increasing when we apply the interpolation.

Reply:
Ah ok, thanks. I'll add a comment explaining that.